### PR TITLE
Joe

### PR DESCRIPTION
**Pull request by Jdonerme**
Added in some link functions

### CODE REVIEW

File: link.py [hunk 1]
```diff
@@ -91,27 +90,61 @@ def cur_destination(self, value):
...
     
     
-    def add_packet(self, packet, host_id):
+    def add_packet(self, packet, node_id):
         """Add a packet to be sent
         
-        Puts the packet in the packet buffer to be sent to to the host with the
-        other id. If no packet is being sent, calls the start_packet_transfer
+        Puts the packet in the packet buffer to be sent to to the node with the
+        other id. If no packet is being sent, calls the start_packet_transfer.
```
Comment (malavika1) [inline link.py]: calls the start_packet_transfer method? don't leave me hanging lol

File: link.py [hunk 2]
```diff
@@ -91,27 +90,61 @@ def cur_destination(self, value):
...
+        time_to_pass = self.cur_packet.packet_size / self.capacity
+        time_to_pass += self.prop_delay
+        
+        event = lambda: self.finish_packet_transfer()
+        print "Link %s beginning to transfer packet %s" \
+              %(self.link_id, self.cur_packet.packet_id)
+        self.ns.add_event(event, time_to_pass)
    
     def finish_packet_transfer(self):
         """Hand off the packet it to the Node it was going to.
```
Comment (malavika1) [inline link.py]: lowercase node

File: link.py [hunk 3]
```diff
@@ -91,27 +90,61 @@ def cur_destination(self, value):
...
+            self.cur_buffer_size += packet.packet_size
+            
+            if self.cur_packet == None:
+                event = lambda: self.start_packet_transfer()
+                self.ns.add_event(event)
+        else:
+            print "Link with id %s is full so packet with \
+            id %s is dropped" % (self.link_id, packet.packet_id)
     
     def start_packet_trasfer(self):
```
Comment (hatooku) [inline link.py]: trasfer -> mispelling

File: link.py (hunk 1 above)
Comment (Jdonerme) [inline link.py]: fixed

File: link.py (hunk 2 above)
Comment (Jdonerme) [inline link.py]: fixed

File: link.py [hunk 4]
```diff
@@ -91,35 +90,81 @@ def cur_destination(self, value):
...
         Takes a packet out of the buffer and begins sending it to the correct
-        host. Sets the current packet and current destination of the link. 
+        node. Sets the current packet and current destination of the link. 
         
         """
+        assert self.cur_destination == None
+        assert self.cur_packet == None 
+        assert len(self.link_buffer) > 0   
+        
+        packet_info = self.link_buffer.popleft()
```
Comment (hatooku) [inline link.py]: This can be done in 1 line.

File: link.py [hunk 5]
```diff
@@ -5,14 +5,13 @@ class Link(object):
        
     Attributes:
         ns (NetworkSimulator): stores the simulator class running the simulation
-        link_id (int): a unique identifier for each link
+        link_id (string): a unique identifier for each link
         max_buffer_size (int): how many bits can be stored in the link buffer
-        prop_delay (float): propogation delay of the link in ms
-        capacity (float): the maximum link rate in bits / ms
+        cur_buffer_size (int): how many bits are in the link buffer
+        prop_delay (float): propogation delay of the link in s
```
Comment (malavika1) [inline link.py]: propagation

File: link.py (hunk 4 above)
Comment (Jdonerme) [inline link.py]: fixed

File: link.py (hunk 3 above)
Comment (Jdonerme) [inline link.py]: fixed

File: link.py [hunk 6]
```diff
@@ -5,14 +5,13 @@ class Link(object):
        
     Attributes:
         ns (NetworkSimulator): stores the simulator class running the simulation
-        link_id (int): a unique identifier for each link
+        link_id (string): a unique identifier for each link
         max_buffer_size (int): how many bits can be stored in the link buffer
```
Comment (mortonrj) [inline link.py]: So in the flow stuff we're making everything ints (even bits).  Idk if you want to do that here for consistency?

File: link.py [hunk 7]
```diff
@@ -91,35 +90,79 @@ def cur_destination(self, value):
...
         When this method is called, the packet will be transfered to the 
         correct node. If there is another packet in the buffer that needs to be
         sent, the start_packet_transfer function will be called and another
         packet will be sent.
         
-        """  
-        pass
\ No newline at end of file
+        """ 
+        event = lambda: self.cur_destination.recieve_packet(self.cur_packet)
```
Comment (hatooku) [inline link.py]: receive, not recieve

File: link.py [hunk 8]
```diff
@@ -5,14 +5,13 @@ class Link(object):
        
     Attributes:
         ns (NetworkSimulator): stores the simulator class running the simulation
-        link_id (int): a unique identifier for each link
+        link_id (string): a unique identifier for each link
         max_buffer_size (int): how many bits can be stored in the link buffer
-        prop_delay (float): propogation delay of the link in ms
-        capacity (float): the maximum link rate in bits / ms
+        cur_buffer_size (int): how many bits are in the link buffer
```
Comment (mortonrj) [inline link.py]: Same

File: link.py (hunk 6 above)
Comment (Jdonerme) [inline link.py]: sounds cute. fixed

File: link.py (hunk 8 above)
Comment (Jdonerme) [inline link.py]: same

**Pull request merged**